lib/ld-eventsource/client.rb

```diff
@@ -163,12 +183,56 @@ def on_error(&action)
     #
     def close
       if @stopped.make_true
-        @cxn.close if !@cxn.nil?
-        @cxn = nil
+        reset_http
       end
     end
 
     private
+    
+    def reset_http
+      @http_client.close if !@http_client.nil?
+      @cxn = nil
+      @buffer = ""
+    end
+    
+    def read_lines
+      Enumerator.new do |gen|
+        loop do
+          line = read_line
+          break if line.nil?
+          gen.yield line
+        end
+      end
+    end
+    
+    def read_line
+      loop do
+        @lock.synchronize do
+          i = @buffer.index(/[\r\n]/)
+          if !i.nil?
+            i += 1 if (@buffer[i] == "\r" && i < @buffer.length - 1 && @buffer[i + 1] == "\n")
```

Comment:
I think there's a flaw here that was also present in my earlier code. We need some special handling here for the edge case where the _very last character in the buffer_ is `\r`. If that happens, we can still consider it to be a line ending— but we need to remember that it happened, and if the very next character we see (after reading more data) is `\n`, then we should skip that and _not_ treat it as a line ending. This is because in SSE, `\n`, `\r`, and `\r\n` are all valid line endings, and we do not want to treat `\r\n` as _two_ line endings in a row.

Reply:
OK I updated this such that if the last character of the buffer (and `i` is the last character) is `\r` that it will not dispatch a line, and instead read more data. If `i` does reference an `\r` with an additional character after it, it will still dispatch.